File: tools/test-personality-options-fixture.js
'use strict';
const assert = require('node:assert/strict');
const personality = require('../formation-personality.js');
const names = personality.PERSONALITY_NAMES;
for (const count of [2, 3, 4, 5]) {
  const basic = { id: `fixture-${count}`, personality: '裏面', personalityOptions: names.slice(0, count) };
  const empty = personality.resolveFormationPersonality(basic, null);
  assert.equal(empty.isSelectable, true);
  assert.equal(empty.needsSelection, true);
  assert.equal(empty.effectivePersonality, null);
  for (const option of basic.personalityOptions) {
    const selected = personality.resolveFormationPersonality(basic, option);
    assert.equal(selected.effectivePersonality, option);
    assert.equal(selected.needsSelection, false);
  }
  if (count < 5) {
    const invalid = personality.resolveFormationPersonality(basic, names[count]);
    assert.equal(invalid.invalidSelection, true);
    assert.equal(invalid.originalSelection, names[count]);
    assert.equal(invalid.effectivePersonality, null);
  }
}
assert.equal(personality.resolveFormationPersonality('純粋', null).effectivePersonality, '純粋');
assert.deepEqual(personality.getPersonalityOptions('共鳴'), names);
const joanne = { id: 'Joanne', personality: '裏面', personalityOptions: ['憂鬱', '純粋'] };
assert.equal(personality.resolveFormationPersonality(joanne, '憂鬱').effectivePersonality, '憂鬱');
assert.equal(personality.resolveFormationPersonality(joanne, '純粋').effectivePersonality, '純粋');
assert.equal(personality.resolveFormationPersonality(joanne, '冷静').invalidSelection, true);
for (const options of [['純粋', '純粋'], ['純粋', '未知'], ['純粋'], names.concat('純粋')]) {
  assert.throws(() => personality.getPersonalityOptions({ personality: '裏面', personalityOptions: options }));
}
assert.throws(() => personality.getPersonalityOptions({ personality: '純粋', personalityOptions: ['純粋', '憂鬱'] }));
for (const base of ['裏面', '未知分類']) {
  for (const value of [undefined, null, '', []]) {
    const row = { id: 'Joanne', personality: base };
    if (value !== undefined) row.personalityOptions = value;
    assert.throws(() => personality.resolveFormationPersonality(row, null), /性格候補/);
  }
}
assert.deepEqual(personality.getPersonalityOptions({}), []);
assert.equal(personality.normalizeStoredSelection(' 冷静 '), '冷静');
assert.equal(personality.normalizeStoredSelection({ personality: '冷静' }), null);
assert.equal(personality.resolveFormationPersonality(joanne, { personality: '純粋' }).effectivePersonality, null);
assert.throws(() => personality.resolveFormationPersonality({ id: 'broken', personality: '' }, null), /性格候補/);
console.log('personality options fixture: OK');

File: tools/test-personality-options-fixture.py
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parent))
from personality_options import normalize_personality_options

for count in (2, 3, 4, 5):
    options = ["純粋", "冷静", "狂気", "活発", "憂鬱"][:count]
    row = normalize_personality_options({"id": f"fixture-{count}", "性格": "裏面", "性格候補": ",".join(options)})
    assert row["personalityOptions"] == options
    assert "性格候補" not in row
assert "personalityOptions" not in normalize_personality_options({"性格": "純粋"})
for raw in ("純粋,純粋", "純粋,未知", "純粋", "純粋,冷静,狂気,活発,憂鬱,純粋"):
    try:
        normalize_personality_options({"性格": "裏面", "性格候補": raw})
        raise AssertionError(raw)
    except ValueError:
        pass
try:
    normalize_personality_options({"性格": "純粋", "性格候補": "純粋,憂鬱"})
    raise AssertionError("fixed personality accepted")
except ValueError:
    pass
for base in ("裏面", "未知分類", ""):
    for raw in (None, ""):
        row = {"id": "Joanne", "性格": base}
        if raw is not None:
            row["性格候補"] = raw
        try:
            normalize_personality_options(row)
            raise AssertionError((base, raw))
        except ValueError as error:
            assert "Joanne" in str(error) and "性格候補" in str(error)
assert "personalityOptions" not in normalize_personality_options({"id": "resonance", "性格": "共鳴"})
print("generator personality options fixture: OK")
